File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/lang/KoreanLanguagePack.kt
package com.draftright.keyboard.lang

import com.draftright.keyboard.Composer
import com.draftright.keyboard.KeyDef
import com.draftright.keyboard.LanguagePack
import com.draftright.keyboard.SpecialKeys
import com.draftright.keyboard.composer.HangulComposer
import java.util.Locale

private fun jamo(vararg labels: String): List<KeyDef> =
    labels.map { KeyDef(it, it[0].code) }

/**
 * Korean (한국어) pack — dubeolsik (2-set) jamo layout. Keys carry the jamo;
 * [HangulComposer] assembles them into syllable blocks live. No candidate bar
 * (Korean is deterministic composition, not dictionary conversion).
 *
 * Long-press a basic consonant/vowel for its tense/extra form
 * (ㅂ→ㅃ, ㄱ→ㄲ, ㅐ→ㅒ, …) — the dubeolsik "shift" set.
 */
object KoreanLanguagePack : LanguagePack {
    override val id = "ko"
    override val displayName = "한국어"
    override val locale: Locale = Locale.KOREAN

    override val alphaRows: List<List<KeyDef>> = listOf(
        jamo("ㅂ", "ㅈ", "ㄷ", "ㄱ", "ㅅ", "ㅛ", "ㅕ", "ㅑ", "ㅐ", "ㅔ"),
        jamo("ㅁ", "ㄴ", "ㅇ", "ㄹ", "ㅎ", "ㅗ", "ㅓ", "ㅏ", "ㅣ"),
        listOf(KeyDef("⬆", SpecialKeys.SHIFT, 1.5f)) +
            jamo("ㅋ", "ㅌ", "ㅊ", "ㅍ", "ㅠ", "ㅜ", "ㅡ") +
            listOf(KeyDef("←", SpecialKeys.BACKSPACE, 1.5f)),
        listOf(
            KeyDef("?123", SpecialKeys.SYMBOLS, 1.5f),
            KeyDef("🌐", SpecialKeys.GLOBE, 1.0f),
            KeyDef(",", ','.code, 1.0f),
            KeyDef(" ", ' '.code, 4.0f),
            KeyDef(".", '.'.code, 1.0f),
            KeyDef("↵", SpecialKeys.ENTER, 1.5f),
        ),
    )

    // Symbols layers shared with the Latin/QWERTY layout.
    override val symbols1Rows: List<List<KeyDef>> = QwertyLayout.symbols1Rows
    override val symbols2Rows: List<List<KeyDef>> = QwertyLayout.symbols2Rows

    override val longPressAccents: Map<Char, List<Char>> = mapOf(
        'ㅂ' to listOf('ㅃ'), 'ㅈ' to listOf('ㅉ'), 'ㄷ' to listOf('ㄸ'),
        'ㄱ' to listOf('ㄲ'), 'ㅅ' to listOf('ㅆ'),
        'ㅐ' to listOf('ㅒ'), 'ㅔ' to listOf('ㅖ'),
    )

    override fun composer(): Composer = HangulComposer()
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Lang/KoreanLanguagePack.swift
import Foundation

/// Korean (한국어) pack — dubeolsik (2-set) jamo layout. Keys carry the jamo;
/// `HangulComposer` assembles them into syllable blocks live. No candidate bar
/// (Korean is deterministic composition). Mirrors the Kotlin KoreanLanguagePack.
///
/// Long-press a basic consonant/vowel for its tense/extra form (ㅂ→ㅃ, ㄱ→ㄲ, …).
public struct KoreanLanguagePack: LanguagePack {
    public let id = "ko"
    public let displayName = "한국어"
    public let locale = Locale(identifier: "ko")

    public let longPressAccents: [Character: [Character]] = [
        "ㅂ": ["ㅃ"], "ㅈ": ["ㅉ"], "ㄷ": ["ㄸ"], "ㄱ": ["ㄲ"], "ㅅ": ["ㅆ"],
        "ㅐ": ["ㅒ"], "ㅔ": ["ㅖ"],
    ]

    public init() {}

    public func makeComposer() -> Composer? { HangulComposer() }

    private func jamo(_ labels: String...) -> [KeyDef] {
        labels.map { KeyDef($0, keyCode($0)) }
    }

    public var alphaRows: [[KeyDef]] {
        [
            jamo("ㅂ", "ㅈ", "ㄷ", "ㄱ", "ㅅ", "ㅛ", "ㅕ", "ㅑ", "ㅐ", "ㅔ"),
            jamo("ㅁ", "ㄴ", "ㅇ", "ㄹ", "ㅎ", "ㅗ", "ㅓ", "ㅏ", "ㅣ"),
            [KeyDef("⬆", SpecialKeys.shift, widthWeight: 1.5)]
                + jamo("ㅋ", "ㅌ", "ㅊ", "ㅍ", "ㅠ", "ㅜ", "ㅡ")
                + [KeyDef("←", SpecialKeys.backspace, widthWeight: 1.5)],
            [
                KeyDef("?123", SpecialKeys.symbols, widthWeight: 1.5),
                KeyDef("🌐", SpecialKeys.globe, widthWeight: 1.0),
                KeyDef(",", keyCode(","), widthWeight: 1.0),
                KeyDef(" ", keyCode(" "), widthWeight: 4.0),
                KeyDef(".", keyCode("."), widthWeight: 1.0),
                KeyDef("↵", SpecialKeys.enter, widthWeight: 1.5),
            ],
        ]
    }

    public var symbols1Rows: [[KeyDef]] { QwertyLayout.symbols1Rows }
    public var symbols2Rows: [[KeyDef]] { QwertyLayout.symbols2Rows }
}
